test/testscript.go

```diff
@@ -0,0 +1,155 @@
+	setupFunc := func(env *testscript.Env) error {
+		src := filepath.Join(e.projectDir, srcRelPath)
+		dst := filepath.Join(env.WorkDir, srcRelPath)
+		dstDir := filepath.Dir(dst)
+		if err := os.MkdirAll(dstDir, 0755); err != nil {
+			msg := "failed to create destination directory %s: %w"
+			return fmt.Errorf(msg, dstDir, err)
+		}
+
+		srcContent, err := os.ReadFile(src)
+		if err != nil {
+			return fmt.Errorf("failed to read source file %s: %w", src, err)
+		}
+
+		if err := os.WriteFile(dst, srcContent, 0644); err != nil {
+			msg := "failed to write destination file %s: %w"
+			return fmt.Errorf(msg, dst, err)
+		}
+		return nil
+	}
+	e.setupFuncs = append(e.setupFuncs, setupFunc)
+	return e
+}
+
+func (e *ProjectTest) RenderTemplateFileFromEnvWithCleanup(
+	srcRelPath string,
+	envKeys []string,
+) *ProjectTest {
+	src := filepath.Join(e.projectDir, srcRelPath)
+	tmpl := template.Must(template.ParseFiles(src))
+	dstRelPath := srcRelPath
+	return e.RenderTemplateFromEnvWithCleanup(tmpl, dstRelPath, envKeys)
+}
+
+func (e *ProjectTest) RenderTemplateStringFromEnvWithCleanup(
+	templateString string,
+	dstRelPath string,
+	envKeys []string,
+) *ProjectTest {
+	tmpl := template.Must(
+		template.New(filepath.Base(dstRelPath)).
+			Parse(templateString),
+	)
+	return e.RenderTemplateFromEnvWithCleanup(tmpl, dstRelPath, envKeys)
+}
+
+func (e *ProjectTest) RenderTemplateFromEnvWithCleanup(
+	tmpl *template.Template,
+	dstRelPath string,
+	envKeys []string,
+) *ProjectTest {
+	setupFunc := func(env *testscript.Env) error {
+		dst := filepath.Join(env.WorkDir, dstRelPath)
+
+		dstDir := filepath.Dir(dst)
+		if err := os.MkdirAll(dstDir, 0755); err != nil {
+			msg := "failed to create destination directory %s: %w"
+			return fmt.Errorf(msg, dstDir, err)
+		}
+
+		envMap := make(map[string]string, len(envKeys))
+		for _, key := range envKeys {
+			val := os.Getenv(key)
+			if val == "" {
+				return fmt.Errorf("environment variable %s is not set", key)
+			}
+			envMap[key] = val
+		}
+
+		dstFile, err := os.Create(dst)
+		if err != nil {
+			return fmt.Errorf("failed to create destination file %s: %w", dst, err)
+		}
+		defer dstFile.Close()
+
+		tmpl.Option("missingkey=error")
+		if err := tmpl.Execute(dstFile, envMap); err != nil {
+			return fmt.Errorf("failed to execute template %s: %v", tmpl.Name(), err)
+		}
+
+		// rendered files may contain secrets, register a cleanup func for removal
+		env.Defer(func() {
+			if err := os.Remove(dst); err != nil {
+				e.t.Fatalf("failed to remove file %s: %v", dst, err)
+			}
+		})
+		return nil
+	}
+	e.setupFuncs = append(e.setupFuncs, setupFunc)
+	return e
+}
+
+func (e *ProjectTest) ExecuteMakeTarget(target string) *ProjectTest {
+	setupFunc := func(env *testscript.Env) error {
+		cmd := exec.Command("make", target)
+		cmd.Dir = e.projectDir
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		e.t.Logf("Running command: '%s'", cmd.String())
```

Comment:
I added a log here so the user understands that a command is being run as part of the test. This is used to run `make build` in the example projects.

Reply:
The one downside to `testscript` is that the output of `go test` is pretty cluttered. The statement you added here does not get printed because our Makefile omits the `-v` flag for `go test`. I added it back in to check and while I can see your log statement now there is also a bunch of output from `testscript`.

Reply:
yeah, the testscript package hooks into the go test framework so it is verbose if -v is used in `go test`, and testscript test output is quite verbose. So far I haven't figured out how to make the testscript output only "slightly verbose". I figure at the very least go switches to verbose output when things fail, so this log will at least be in there if the make command happens to be part of the problem. Open to suggestions/improvements for dev usability w.r.t. test output

Reply:
keeping as is for now!